include/beast/core/handler_ptr.hpp

```diff
@@ -0,0 +1,177 @@
+//
+// Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_HANDLER_PTR_HPP
+#define BEAST_HANDLER_PTR_HPP
+
+#include <beast/core/detail/type_traits.hpp>
+#include <atomic>
+#include <cstdint>
+#include <utility>
+
+namespace beast {
+
+namespace detail {
+
+template<class T, class = beast::detail::void_t<>>
+struct has_handler_type : std::false_type {};
+
+template<class T>
+struct has_handler_type<T, beast::detail::void_t<
+    typename T::handler_type
+        > > : std::true_type {};
+
+} // detail
+
+/** A smart pointer container.
+
+    This is a smart pointer that retains shared ownership of an
+    object through a pointer. Memory is managed using the allocation
+    and deallocation functions associated with a completion handler,
+    which is also stored in the object. The object is destroyed and
+    its memory deallocated when one of the following happens:
+
+    * The function @ref invoke is called.
+    * The function @release_handler is called
```

Comment:
Sorry Vinnie, should be `@ref release_handler`.

The Markdown-style `*` bullets should be `@li` instead, I think. They're not rendered correctly in the docs: http://vinniefalco.github.io/stage/beast/b22/beast/ref/handler_ptr.html